### PR TITLE
Markdown reporter

### PR DESCRIPTION
**Pull request by Gerrrr**
Solution for https://github.com/FRosner/drunken-data-quality/issues/40

### CODE REVIEW

**Comment by FRosner** [comment]
I sense that the pull request is malicious... :dart:

**Comment by FRosner** [comment]
Thanks for the fixes @Gerrrr. Is it now ready to be looked at or is there something you would like to change?

**Comment by Gerrrr** [comment]
Yes, it is now ready to be looked. Looking forward to hear your feedback.

**Comment by FRosner** [comment]
Here are my remarks. I will change them on my own and merge the branch afterwards. So no need to update the code, you can just reply to this conversation.
1. Test package structure in `test` should be mirrored from `main`. E.g. the `ReporterTest` should also be in `de.frosner.ddq.reporters`.
2. I would like to have a few formattings in a different way (will make separate commits for each of them so we can discuss if necessary).
3. We should try to separate a bit more between unit and integration tests. E.g. for testing a simple report function of a check result, we should not be pulling in a Spark context to create a data frame. However, I think having a few integration tests is nice because it tests all stuff together and also documents the usage quite nicely.

**Comment by FRosner** [comment]
Could you please provide a compiled Markdown example report? Maybe having a few success and failure messages? You can just copy paste here, because the comments should support Markdown anyway.

File: src/test/scala/de/frosner/ddq/CheckTest.scala [hunk 1]
```diff
@@ -36,55 +36,72 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
     sql.createDataFrame(rdd, schema)
   }
 
-  "A number of rows equality check" should "succeed if the number of rows is equal to the expected" in {
-    Check(makeIntegerDf(List(1, 2, 3))).hasNumRowsEqualTo(3).run shouldBe true
-  }
-
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/CheckTest.scala:42]: @Gerrrr why did you delete this one?

File: src/test/scala/de/frosner/ddq/CheckTest.scala [hunk 2]
```diff
@@ -296,7 +430,9 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
     val df = makeIntegerDf(List(1,2,3), sql)
     val tableName = "myintegerdf1"
     df.registerTempTable(tableName)
-    Check.sqlTable(sql, tableName).hasNumRowsEqualTo(3).run shouldBe true
+    val constraint = Check.hasNumRowsEqualTo(3)
+    val result = ConstraintSuccess("The number of rows is equal to 3")
+    Check.sqlTable(sql, tableName).addConstraint(constraint).run().constraintResults shouldBe Map(constraint -> result)
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/CheckTest.scala]: @Gerrrr Actually, this test can now be improved. It does no actually test the table name output so I will either add this to the `RunnerTest` or create a new integration test that checks if the table name is put as a header in the `CheckResult`

File: src/test/scala/de/frosner/ddq/CheckTest.scala [hunk 3]
```diff
@@ -120,172 +146,280 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    val constraint3 = check.constraints(2)
+
+    check.run().constraintResults shouldBe Map(
+      constraint1 -> ConstraintSuccess("The number of rows is equal to 3"),
+      constraint2 -> ConstraintFailure("The actual number of rows 3 is not equal to the expected 2"),
+      constraint3 -> ConstraintSuccess("Constraint column > 0 is satisfied")
+    )
   }
 
   "It" should "be possible to specify a display name for a data frame" in {
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/CheckTest.scala:425]: @Gerrrr Actually, this test can now be improved. It does no actually test the table name output so I will either add this to the `RunnerTest` or create a new integration test that checks if the table name is put as a header in the `CheckResult`

**Comment by Gerrrr** [comment]
Markdown example report:

# Checking [Outlook: string, Temperature: string, Humidity: string, Wind: string, Play: string]

It has a total number of 5 columns and 14 rows.
- [success]: Column Wind can be converted to Boolean
- [failure]: Column Outlook contains 9 rows that are not in Set(sunny)
- [success]: Column Temperature can be converted to Double
- [failure]: Column Play contains 5 rows that are not in Set(yes)
- [success]: The number of rows is equal to 14

File: src/test/scala/de/frosner/ddq/CheckTest.scala (hunk 1 above)
Comment (Gerrrr) [inline src/test/scala/de/frosner/ddq/CheckTest.scala:42]: @FRosner it was deleted by accident.

**Comment by FRosner** [comment]
@Gerrrr thanks for the markdown example. Two things I would like to raise:
1. The heading might be a bit too heavy. However, I don't know what an appropriate heading would be. Another option is that we just make it bold.
2. Success and failure statements don't look very appealing. First of all you have an enumeration but then you also have success or failure. So actually it should be enough to enumerate with success of failure. Also, some formatting could be nice.

Please find some ideas on how it could also look below.

---

**Checking [Outlook: string, Temperature: string, Humidity: string, Wind: string, Play: string]**

It has a total number of 5 columns and 14 rows.
- _SUCCESS_: Column Wind can be converted to Boolean
- _FAILURE_: Column Outlook contains 9 rows that are not in Set(sunny)

---

**Checking [Outlook: string, Temperature: string, Humidity: string, Wind: string, Play: string]**

It has a total number of 5 columns and 14 rows.
- :white_check_mark: Column Wind can be converted to Boolean
- :x: Column Outlook contains 9 rows that are not in Set(sunny)

File: src/test/scala/de/frosner/ddq/RunnerTest.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    val checkResults = Runner.run(List(check1, check2), List())
+
+    checkResults.size shouldBe 2
+
+    val checkResult1 = checkResults.head
+    val checkResult2 = checkResults.drop(1).head
+
+    checkResult1.check shouldBe check1
+    checkResult1.constraintResults shouldBe Map((constraint1, result1))
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/RunnerTest.scala:35]: I think we should also add expectations for header and prologue here. I will do it.

**Comment by FRosner** [comment]
We also need to test if the persist method if used correctly in the run method. I will add this test case as well.

**Comment by FRosner** [comment]
I am done with my stuff and created branch issue/40 that is ready for merge. As soon as the Travis and Coveralls gives green light, I am going to merge it and close this one.
